constantine/signatures/ecdsa.nim

```diff
@@ -0,0 +1,311 @@
+
+import std / macros # for `update` convenience helper
+
+export ecc_sig_ops
+
+type
+  ## Decides the type of sampler we use for the nonce. By default
+  ## a simple uniform random sampler. Alternatively a deterministic
+  ## sampler based on message hash and private key.
+  NonceSampler* = enum
+    nsRandom, ## pure uniform random sampling
+    nsRfc6979 ## deterministic according to RFC 6979
+
+proc toBytes[Name: static Algebra; N: static int](res: var array[N, byte], x: FF[Name]) =
+  discard res.marshal(x.toBig(), bigEndian)
+
+func fromDigest[Name: static Algebra; N: static int](
+    dst: var Fr[Name], src: array[N, byte],
+    truncateInput: static bool): bool {.discardable.} =
+  ## Convert a hash function digest to an element in the scalar field `Fr[Name]`.
+  ## The proc returns a boolean indicating whether the data in `src` is
+  ## smaller than the field modulus. It is discardable, because in some
+  ## use cases this is fine (e.g. constructing a field element from a hash),
+  ## but invalid in the nonce generation following RFC6979.
+  ##
+  ## The `truncateInput` argument handles how `src` arrays larger than the BigInt
+  ## underlying `Fr[Name]` are handled. If it is `false` we will simply throw
+  ## an assertion error on `unmarshal` (used in RFC6979 nonce generation where
+  ## the array size cannot be larger than `Fr[Name]`). If it is `true`, we truncate
+  ## the digest array to the left most bits of up to the number of bits underlying
+  ## the BigInt of `Fr[Name]` following SEC1v2 [0] (page 45, 5.1-5.4).
+  ##
+  ## [0]: https://www.secg.org/sec1-v2.pdf
+  var scalar {.noInit.}: matchingOrderBigInt(Name)
+  when truncateInput: # for signature & verification
+    # If the `src` array is larger than the BigInt underlying `Fr[Name]`, need
+    # to truncate the `src`.
+    const OctetWidth = 8
+    const FrBytes = Fr[Name].bits.ceildiv_vartime(OctetWidth)
+    # effectively: `scalar ~ array[0 ..< scalar.len]`
+    scalar.unmarshal(toOpenArray[byte](src, 0, FrBytes-1), bigEndian)
+    # Now still need to right shift potential individual bits.
+    # e.g. 381 bit BigInt fits into 384 bit (48 bytes), so need to
+    # right shift 3 bits to truncate correctly.
+    const toShift = FrBytes * OctetWidth - Fr[Name].bits
+    when toShift > 0:
+      scalar.shiftRight(toShift)
+  else: # for RFC 6979 nonce sampling. If larger than modulus, sample again
+    scalar.unmarshal(src, bigEndian)
+  scalar.unmarshal(src, bigEndian)
+  # `true` if smaller than modulus
+  result = bool(scalar < Fr[Name].getModulus())
+  dst.fromBig(scalar)
+
+proc randomFieldElement[FF](): FF =
+  ## random element in ~Fp[T]/Fr[T]~
+  let m = FF.getModulus()
+  var b: matchingBigInt(FF.Name)
+
+  while b.isZero().bool or (b > m).bool:
+    ## XXX: raise / what else to do if `sysrand` call fails?
+    doAssert b.limbs.sysrand()
+
+  result.fromBig(b)
+
+proc arrayWith[N: static int](res: var array[N, byte], val: byte) =
```

Comment:
```suggestion
proc byteArrayWith(N: static int, val: byte): array[N, byte] {.noInit, inline.} =
```

Reply:
This is only used twice and just after initialization. We can avoid bad codegen with `{.inline.}` and we can avoid avoid an extra `memset(0)` with `{.noInit.}`